### ISSUE TITLE
`dontCacheBustUrlsMatching` ignored in Workbox v3?

### ISSUE THREAD

**Issue by freezy** [GoogleChrome/workbox#1373]
**Library Affected**:
workbox-webpack-plugin

**Browser & Platform**:
Webpack 4.1.1, Workbox 3.0.0, Node 8.9.1, Windows 10 x64

**Issue or Feature Request Description**:
When migrating to v3 I noticed that most of my assets are cache-busted when they used not to be with v2.

You can reproduce using this repo:

```bash
$ git clone https://github.com/vpdb/website
$ cd website && git checkout workbox3
$ npm install && npm run build:prod
```
For example, the manifest contains:
```js
{
    "revision": "ad84f24427599d1c1b955914cfd9a4fa",
    "url": "https://vpdb.io/profile/settings/profile.settings-a1VkNzRv.html"
}
```
whereas v2 of Workbox produces:
```js
{
    "url": "https://vpdb.io/profile/settings/profile.settings-a1VkNzRv.html"
}
```

The [config](https://github.com/vpdb/website/blob/workbox3/config/webpack.prod.js#L26) uses this value:

    dontCacheBustUrlsMatching: /-([0-9a-f]{12}|[0-9A-Za-z]{8})\./,

I've also noticed that `globIgnores` has the same problem, while `exclude` works fine. Furthermore, chunks never seem to be cache-busted, even when `dontCacheBustUrlsMatching` isn't set.

**Comment by jeffposnick** [comment]
See https://github.com/GoogleChrome/workbox/issues/1380 and https://developers.google.com/web/tools/workbox/modules/workbox-webpack-plugin#cache_additional_non-webpack_assets

We need to do a better job of explaining this to folks.

In general, if you've got `webpack` configured to include hash value in an asset name, the `workbox-webpack-plugin` attempts to detect this automatically and excludes the revision field by default.

The detection of which assets include `[hash]` or `[chunkhash]` or something related to that is unfortunately not 100% accurate. See https://stackoverflow.com/questions/47658134

**Comment by freezy** [comment]
Well the glob thing was just something I noticed. All my assets run through Webpack, so this doesn't seem to be an issue (`exclude` covers this just fine).

When you say you're detecting hashes in assets automatically, does that mean that the `dontCacheBustUrlsMatching` option is now only applied to non-Webpack assets? That would be a breaking change, right?

**Comment by tmchng** [comment]
@jeffposnick 

Why don't `dontCacheBustUrlsMatching` and `manifestTransform` apply to webpack generated files? I think the reason why we are confused is because intuitively they should apply to all items coming out of the webpack pipeline.

For instance, I'm appending **git commit hash** to my webpack-compiled files instead of using `[hash]` and `[chunkhash]`, and I wish to precache these files without revisioning inside the precache-manifest. I don't think there's an option for me to do that right now.

The way that `dontCacheBustUrlsMatching` and `manifestTransform` currently works feels like an undesired feature (or a bug). I think it makes a lot more sense for them to apply to ALL items that are about to get written into the manifest.

I'd be more than happy to do a PR for this change. Please let me know what you think, thanks!

**Comment by freezy** [comment]
There is also #1458 which deals with the hash length, but the problem is that I'm using different hash lengths depending on the encoding (hex gets more chars than base58). 

Getting `dontCacheBustUrlsMatching` back for Webpack assets would really be ideal here.

**Comment by jeffposnick** [comment]
First, some additional background info. Apologies if this is repetitive or obvious:

There's a general tension between the two sources of URLs that might be precached when using the `webpack` plugins: URLs that originate from the `webpack` asset pipeline, and URLs that exist outside of the `webpack` build process (i.e. files that exist on disk and which are matches via `glob` patterns).

In Workbox v2, we didn't have that tension. We picked up everything via `glob` patterns, and did not attempt to take advantage of the context provided by `webpack`. We assumed that `dontCacheBustUrlsMatching` or `manifestTransforms`, if provided, would be apply to anything picked up during the build process, including `webpack` assets.

In Workbox v3, we moved to a model that attempted to infer as much context as possible from the `webpack` build process. Some examples of this are detecting whether a URL already contains a hash, or picking up the `publicPath` configuration rather than requiring a `manifestTransform` to explicitly adjust URLs.

At the same time, because we know that some users *still* needed to precache additional files that exist outside of `webpack`, we allowed them to continue setting `dontCacheBustUrlsMatching`/`manifestTransform`/etc., and only apply them to the `glob`-derived precache manifest, matching the v2 behavior.

We may need to swing back towards the other direction now, but the cleanest approach is up for debate.

One approach would be to just take all the options that only apply to `glob` matches in the v3 `webpack` plugin, and apply them to `webpack` assets as well. The drawback here is that while some options are probably harmless (`dontCacheBustUrlsMatching`, for instance, is unlikely to trigger too many false positives), other options, like a `manifestTransform` which modifies a URL prefix, might really only make sense when used against `glob` matches and not against `webpack` assets. There's the potential for going this route and trying to infer intent—if `manifestTransforms` is provided *and* there's not `globPatterns`, then that's a signal that the developer really does want their `webpack` assets-derived manifest transformed.

Another approach would be to create a new subsection of the options that applies only to `webpack` assets, which you could then use to unambiguously state that you're cool with those specific modifications being applied to the precache manifest that's tied to `webpack`.
```

And then there are some hybrid approaches, where we might just start applying a subset of options, like `dontCacheBustUrlsMatchin` to both flavors of precache manifests, and only apply others, like `manifestTransforms`, to `glob`-derived manifests.

My personal opinion is that we can provide the best `webpack` developer experience by inferring as much as possible based on the `webpack` build metadata, but this feedback from the community is helpful as we gauge which escape hatches to add for developers who really need it. Anything that amounts to a  breaking change would only happen as part of a major version bump, and in the meantime, we may pursue some tactical fixes that could maintain backwards compatibility.

**Comment by freezy** [comment]
Thanks for providing more background. In order not to break BC, I would find it most intuitive to just get another `dontCacheBustWebpackUrlsMatching` option that applies to Webpack assets.

In the future, grouping common options between globs and assets would probably be a better approach, which would make it clearer which option belongs to which. You'll probably want to find a term that is more explicit, since "globbing" might gets confused with simply selecting stuff from the webpack assets.

**Comment by jeffposnick** [comment]
C.f. https://github.com/webpack/webpack/issues/9038

**Comment by jeffposnick** [comment]
`dontCacheBustURLsMatching` will be added to the webpack plugin in v5, and will apply only to webpack asset URLs.

**Comment by jeffposnick** [comment]
This should be addressed by the current Workbox v5.0.0 alpha.
